Fix mypy errors in testcommon py

HawkinsBA opened the pull request:
- [x] xref [#28926](https://github.com/pandas-dev/pandas/issues/28926)
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

jbrockmendel committed:
BUG: Fix TypeError in _cython_agg_blocks (#29035)

jbrockmendel committed:
CI: xfail on numpy 1.18 (#29057)

* xfail on numpy 1.18

* CI: try using numpy wheel

jbrockmendel committed:
Stop catching TypeError in groupby methods (#29060)

TomAugspurger committed:
API: Restore getting name from MultiIndex level (#29061)

* API: Restore getting name from MultiIndex level

xref https://issues.apache.org/jira/browse/ARROW-6922 /
https://github.com/pandas-dev/pandas/pull/27242#issuecomment-543302582
/ https://github.com/pandas-dev/pandas/issues/29032

No docs yet, since it isn't clear how this will eventually sort out. But
we at least want to preserve this behavior for 1.0

* fixups

iamrajhans committed:
fixed issue of mypy for test_ujson (#29022)

jbrockmendel committed:
CLN: derivation of nogil param (#29047)

jbrockmendel committed:
REF: remove groupby_helper (#29040)

moi90 committed:
ENH: Informative dtype message for for assert_series_equal (#28993)

crepererum committed:
TST: add regression test for all-none-groupby (#29067)

Closes #21624

lrm25 committed:
Fix mypy errors for pandas\tests\*:  test_convert_to.py (#28965)

jbrockmendel committed:
CLN: catch less in groupby (#29077)

jbrockmendel committed:
CLN: simplify take_2d_multi (#29065)

jbrockmendel committed:
REF: use fused types in algos_rank_helper (#29044)

javad94 committed:
DOC: updated categorical docstring (#29068)

jbrockmendel committed:
CLN: tighten exception catching in indexes (#29078)

SaturnFromTitan committed:
[#22550] Remove TestData from series-tests test_quantile.py (#29096)

jbrockmendel committed:
comments, catch less (#29088)

jbrockmendel committed:
REF: remove algos_rank_helper (#29086)

jbrockmendel committed:
CLN: Exception in DataFrame._reduce (#29085)

SaturnFromTitan committed:
[#22550] Remove TestData from series-tests test_operators.py (#29084)

jhereth committed:
TST: 2d index when constructing dataframe (#25416). (#29083)

jhereth committed:
TST: regression test for groupby with datetime and timedelta (#15562) (#29063)

jbrockmendel committed:
CLN: cython and docstring cleanups (#29089)

jbrockmendel committed:
BUG: parse_time_string failing to raise TypeError (#29098)

Commit:
Replacing list concatenation with "+"

grigoriosgiann committed:
Add a regression test for the timezone issue (#29097)

AbhijeetKrishnan committed:
Fix typing errors (#29115)

Thanks, @AbhijeetKrishnan

AbhijeetKrishnan committed:
Fix typing errors (#29114)

AbhijeetKrishnan committed:
Fix mypy errors (#29108)

yogendrasoni committed:
fix #28926 mypy error in pandas\tests\arrays\test_array.py  (#28970)

albertvillanova committed:
TST: Test key dtype cast after merge (#29030)

lukasbk committed:
CLN: Type error fix in tests\tseries\offsets\test_yqm_offsets.py (#28996)

lukasbk committed:
CLN: Type error fix in tests\tseries\offsets\test_yqm_offsets.py (#28996)

jjlkant committed:
PERF: Benchmark merge with non-int64 and tolerance (#28922) (#28974)

R1j1t committed:
Pandas get_dummies validate `columns` input (#28463)

leftys committed:
fix Rolling for multi-index and reversed index. (#28297)

AntonioAndraues committed:
BUG: set_precision format fixed (#13257) (#27934)

datapythonista committed:
WEB: Adding new pandas logo (#28948)

simonjayhawkins committed:
CLN: remove versionadded:: 0.19.0 and earlier (#29127)

simonjayhawkins committed:
CLN: remove versionadded:: 0.20 (#29126)

jbrockmendel committed:
BUG: fix AttributeError raised in libreduction (#29100)

jbrockmendel committed:
REF: avoid getattr pattern for rank_1d, rank_2d (#29137)

jbrockmendel committed:
REF: avoid getattr pattern for diff_2d; use fused types (#29120)

jbrockmendel committed:
REF: avoid getattr pattern for unstack, assorted cleanups (#29121)

jbrockmendel committed:
REF: avoid getattr pattern for join_indexer (#29117)

SaturnFromTitan committed:
Remove TestData from series-tests test_replace.py (#29147)

jbrockmendel committed:
TST: stop suppressing plot test exceptions, xfail broken test (#29099)

gorski-m committed:
TST: Add regression test for Series dropping uint datatype (#18311) (#29071)

jbrockmendel committed:
CLN: Exception in NDFrame._maybe_update_cacher (#29066)

TomAugspurger committed:
REF: Store metadata in an attrs dict (#29062)

SaturnFromTitan committed:
Remove TestData from series-tests test_sorting.py (#29149)

SaturnFromTitan committed:
Remove TestData from series-tests test_timeseries.py (#29150)

punndcoder28 committed:
minor inconsistency in Categorical.remove_categories error message (#28677)

SaturnFromTitan committed:
Remove TestData from series-tests test_api.py (#29153)

simonjayhawkins committed:
CLN: replace Dict with Mapping to annotate arguments (#29155)

jgehrcke committed:
DOC: attempt to fix contributors for 0.24.0, 0.25.0/1/2 (#29152)

gfyoung committed:
Remove unnecessary check for Python >= 3.5 (#29167)

jbrockmendel committed:
REF: avoid getattr pattern for groupby-rank functions (#29166)

simonjayhawkins committed:
TYPING : Series.name -> Optional[Hashable] (#29164)

* TYPING : Series.name -> Optional[Hashable]

gfyoung committed:
Clarify referred command for flake8 caveat (#29168)

Commit:
Update test_common.py

Apply black formatting to test_common.py

jjlkant committed:
CLN: Fix typing in pandas\tests\arrays\test_datetimelike.py (#28926) (#29014)

jbrockmendel committed:
BUG: fix TypeError raised in maybe_downcast_numeric (#29103)

jbrockmendel committed:
Stop catching TypeError in _aggregate_item_by_item (#29177)

jbrockmendel committed:
CLN: Assorted cleanups (#29175)

SaturnFromTitan committed:
Remove TestData from series-tests test_repr.py (#29148)

WillAyd commented:
Can you post the error this solves? Does annotating these instead of unpacking change anything?

jorisvandenbossche committed:
BUG/TST: ensure groupby.agg preserves extension dtype (#29144)

HughKelley committed:
Period 29073 (#29159)

gfyoung committed:
TST: Test Series integer mod behavior on string (#29180)

mohitanand001 committed:
To string with encoding (#28951)

TomAugspurger committed:
DOC: Update the LaTeX author. (#29184)

The LaTeX engine did not like the newlines. It caused an
undefined control sequence.

gfyoung committed:
TST: Document and test na_filter in read_excel (#29171)

WillAyd committed:
CI: Avoid timeouts in 38 build (#29185)

HawkinsBA commented:
pandas\tests\dtypes\test_common.py:327: error: Unsupported operand types for + ("List[Series]" and "List[str]")
pandas\tests\dtypes\test_common.py:355: error: Unsupported operand types for + ("List[Series]" and "List[object]")
pandas\tests\dtypes\test_common.py:387: error: Unsupported operand types for + ("List[Series]" and "List[str]")

In the original file it looks like both lists of Strings and Dtypes (ALL_INT_DTYPES,and to_numpy_dtypes(ALL_INT_DTYPES)) are are concatenated to a List[Series], so my rationale behind unpacking these is to preserve both of these list types being concatenated.

Would it be better to approach this problem preserving concatenation with type annotations?

WillAyd commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -324,11 +324,13 @@ def test_is_datetimelike():
-    + ALL_EA_INT_DTYPES
-    + to_ea_dtypes(ALL_EA_INT_DTYPES),
+    [
+        type(pd.Series([1, 2])),
```
Is this change to `type` intentional?

HawkinsBA replied:
Yes. If I understand the list correctly then a Series is a data type that should be accounted for in these tests, right?

WillAyd replied:
I can see where the naming is confusing but you should keep that the same - `is_integer_dtype` works on types and arrays

HawkinsBA replied:
Alright, I reverted that change.

WillAyd reviewed (changes requested):
OK might just have to live with the unpacking then

rhstanton committed:
BUG: Increase range of dates for holiday calculations (#27790)

jbrockmendel committed:
CLN: preempt TypeError for EAs in groupby agg_series (#29186)

jbrockmendel committed:
CLN: dont catch Exception in _aggregate_frame (#29194)

jbrockmendel committed:
CLN: remove Block._try_coerce_arg (#29139)

SaturnFromTitan committed:
Remove TestData from series-tests test_rank.py (#29101)

kaathi524 committed:
CLN: remove noqa E241 #29207 (#29209)

TomAugspurger committed:
API: Change matplotlib formatter registration (#28722)

SaturnFromTitan committed:
Remove TestData in frame tests in multiple files - part2 (#29222)

gfyoung committed:
CLN: Remove unnecessary sys.version_info checks (#29210)

SaturnFromTitan committed:
Remove TestData in frame tests in multiple files  (#29172)

ellequelle committed:
BUG: fix non-existent variable in NDFrame.interpolate (#29142)

jorisvandenbossche committed:
ENH: Add StringArray.__arrow_array__ for conversion to Arrow (#29182)

jbrockmendel committed:
CLN: Add types in a handful of places (#29178)

datapythonista committed:
Changing logo in README and removing old logo (#29215)

proost committed:
ENH: groupby missing data in index (#28097)

gfyoung committed:
TST: Add test for nested JSON normalization (#29225)

JasperVanDenBosch committed:
BUG: fixes formatted value error for missing sheet (#27676) (#27677)

jbrockmendel committed:
CLN: AttributeError in _wrap_applied_output (#29195)

jbrockmendel committed:
CLN: simplify core.algorithms (#29199)

WillAyd committed:
Update apt-get archives (#29230)

Commit:
Update test_common.py

Removed type(pd.Series[...])

WillAyd committed:
Removed raise_from_traceback (#29174)

simonjayhawkins committed:
CLN: remove simplejson (#29169)

qwhelan committed:
BENCH: Improve perf of rolling.Apply.time_rolling (#29239)

jbrockmendel committed:
CLN: remove algorithms.match (#29249)

jbrockmendel commented:
@HawkinsBA can you rebase

srcole committed:
CLN: fix mypy errors in pandas\tests\indexes\interval\test_base.py #28926 (#28961)

Commit:
Merge remote-tracking branch 'upstream/master' into Fix-mypy-errors-in-test-common-py

Commit:
Update setup.cfg

jbrockmendel commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -352,11 +354,13 @@ def test_is_not_integer_dtype(dtype):
-    + SIGNED_INT_DTYPES
-    + to_numpy_dtypes(SIGNED_INT_DTYPES)
-    + SIGNED_EA_INT_DTYPES
-    + to_ea_dtypes(SIGNED_EA_INT_DTYPES),
```
what is mypy's complaint about this?  I think the current idiom is more beginner-friendly than the args-expansion idiom

WillAyd replied:
IIRC the issue is that you can't really do list addition and have it type inference appropriately all in one pass, at least not from a static analysis perspective

Instead of doing this could you also just initialize this a line up with the appropriate type comment and just pass that variable into the parametrization?

HawkinsBA replied:
@WillAyd Just so I understand exactly what you're suggesting, are you talking about initializing the list of dtypes originially constructed by list addition through unpacking them into a list outside of the parametrizations, giving that list a type stub (presumably List[Dtype]), and then passing it into the paramaterization?

WillAyd replied:
No unpacking just defined this as a variable on the line above. 

```python
signed_int_dtypes = []  # type: List[Union[pd.Series, Dtype]]
```

And then just use that as the argument here.  

Side note - I'm assuming the type above works but can't remember what the mypy complaint was and what it should be just glancing at it. Might need to tweak depending on what you see

HawkinsBA replied:
Okay, those variables have been added and given type stubs for the tests that were throwing mypy errors. It looks like list addition is working without any problems now.

Commit:
Update test_common.py

Commit:
Applied black formatting to test_common.py

Commit:
Update test_common.py

Trying to fix Azure Pipelines

Commit:
Update test_common.py

Trying to fix Azure Pipelines again

Commit:
Update test_common.py

WillAyd commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -322,9 +325,13 @@ def test_is_datetimelike():
     assert com.is_datetimelike(s)
 
 
+integer_dtypes = []  # type: List[Union[Series, str, str, Any, ExtensionDtype]
```
You should only need `str` once within the `Union`. `Any` is also a little too loose - can that not be made more specific?

WillAyd replied:
Also I think you can just add all of the elements this should contain here, rather than defining and adding this later

WillAyd commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -382,9 +393,13 @@ def test_is_not_signed_integer_dtype(dtype):
     assert not com.is_signed_integer_dtype(dtype)
 
 
+unsigned_integer_dtypes = []  # type: List[Union[Series, str, str, Any, ExtensionDtype]
```
Same comment here

jbrockmendel commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -13,6 +15,7 @@
 )
 
 import pandas as pd
+from pandas._typing import Dtype, ExtensionDtype
```
ExtensionDtype needs to come from pandas.core.dtypes.dtypes

Commit:
Updating type annotation in test_common.py

Commit:
Remove unused import from test_common.py

Commit:
Trying to fix Azure Pipelines

pep8speaks commented:
Hello @HawkinsBA! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/tests/dtypes/test_common.py`](https://github.com/pandas-dev/pandas/blob/8d21fe0045592586cf135b7781cbffbc81a633a8/pandas/tests/dtypes/test_common.py):

> [Line 361:89](https://github.com/pandas-dev/pandas/blob/8d21fe0045592586cf135b7781cbffbc81a633a8/pandas/tests/dtypes/test_common.py#L361): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)
> [Line 397:89](https://github.com/pandas-dev/pandas/blob/8d21fe0045592586cf135b7781cbffbc81a633a8/pandas/tests/dtypes/test_common.py#L397): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 88 characters)

HawkinsBA commented:
So without `Any` in the type annotation I get the following MyPy errors while the Azure Pipelines run:

`pandas/tests/dtypes/test_common.py:336: error: Unsupported operand types for + ("List[Union[Series, str, str, Any, ExtensionDtype, ExtensionDtype]]" and "List[str]")
pandas/tests/dtypes/test_common.py:404: error: Unsupported operand types for + ("List[Union[Series, str, str, Any, ExtensionDtype, ExtensionDtype]]" and "List[str]")`

However when 'Any' is included in the type annotation I get the following errors:

`pandas/tests/dtypes/test_common.py:336: error: Unsupported operand types for + ("List[Union[Series, str, str, Any, ExtensionDtype, Any, ExtensionDtype]]" and "List[str]")
pandas/tests/dtypes/test_common.py:404: error: Unsupported operand types for + ("List[Union[Series, str, str, Any, ExtensionDtype, Any, ExtensionDtype]]" and "List[str]")`

I'm not sure what to make of this, but I'm looking into trying to refine what the `Any` might be into a concrete data type.

WillAyd commented:
Does changing from `List` to `Sequence` help? Haven't reviewed thoroughly but the invariant nature of the List might be the issue

Commit:
Replacing List with Sequence and removing Any

Commit:
Applying black formatting

WillAyd commented on pandas/tests/dtypes/test_common.py:
```diff
@@ -322,9 +326,13 @@ def test_is_datetimelike():
     assert com.is_datetimelike(s)
 
 
+integer_dtypes = []  # type: Sequence[Union[pd.Series, str, Dtype, ExtensionDtype]]
```
OK I think just `type: List` will be fine for these

Commit:
Replaciing type stubs with List

Commit:
Remove unused import from test_common.py

WillAyd reviewed (approved):
lgtm. @simonjayhawkins care to take a look as well?

@HawkinsBA the CI failure is unrelated. If you merge master and re-push should go away

simonjayhawkins commented:
> lgtm. @simonjayhawkins care to take a look as well?

@WillAyd 

noting @jbrockmendel comment regarding unpacking, which mypy accepts in an expression. I'm not convinced that this is more clear. Especially, since the use of generic List without type parameters is the same as List[Any] so does not add value to the type checking.

I know we have different opinions on the use of type: ignore but could potentially be used here and reference the mypy issue on list concatenation in expressions?

alternatively, perhaps assign all the values to the list in the initial assignment? could still use #type: List but would not need an expression concatenating items to an empty list. I think this would be clearer to a newcomer. Maybe use block capitals for the variable names. wdyt?

WillAyd commented:
I don't have a strong preference; whatever works for me

simonjayhawkins commented:
both these suggestions are effectively silencing mypy, whereas at least with unpacking the inferred type should have been more precise.

@jbrockmendel considering the above, are you against unpacking? This problem is likely to surface again?

WillAyd commented:
Misinterpreted your earlier comment but I would prefer not to change to unpacking. These are just arguments to parameterize; we aren't mapping any kind of operations against the entire container so we don't need to fine-grain inference; either List or `ignore` would be OK I think

simonjayhawkins commented:
I'm OK with just `List` but think the assignment should be in one statement for clarity.

Commit:
Merge remote-tracking branch 'upstream/master' into Fix-mypy-errors-in-test-common-py

jbrockmendel commented:
> considering the above, are you against unpacking? This problem is likely to surface again?

If you're referring to the `*args` pattern I'd consider it near-last-resort, but not a strong enough opinion to hold up a PR if there is consensus.

simonjayhawkins commented:
yeah, so I think we're agreed.

summarising since this (mypy) issue will crop up again...

the option that works for mypy, i.e. unpacking, means changing the code and is not newcomer friendly.

type:ignore may mask other issues

so refactoring out list concatenation in an expression into a variable and adding a variable annotation is the preferred approach.

simonjayhawkins reviewed (approved):
> so refactoring out list concatenation in an expression into a variable and adding a variable annotation is the preferred approach.

it appears that to do this in one statement would require a cast...

```python
integer_dtypes = (
    cast(List, [pd.Series([1, 2])])
    + ALL_INT_DTYPES
    + to_numpy_dtypes(ALL_INT_DTYPES)
    + ALL_EA_INT_DTYPES
    + to_ea_dtypes(ALL_EA_INT_DTYPES)
)


@pytest.mark.parametrize("dtype", integer_dtypes)
```

otherwise would still need two statements...

```python
integer_dtypes = []  # type: List
integer_dtypes = (
    integer_dtypes
    + [pd.Series([1, 2])]
    + ALL_INT_DTYPES
    + to_numpy_dtypes(ALL_INT_DTYPES)
    + ALL_EA_INT_DTYPES
    + to_ea_dtypes(ALL_EA_INT_DTYPES)
)


@pytest.mark.parametrize("dtype", integer_dtypes)
```

@WillAyd given that this is still not much clearer, merge as is if happy.

WillAyd commented:
Yea pretty tricky with these...I suppose the real intent is tough to realize from a static analysis. Curious to see how many we come across going forward

WillAyd commented:
Thanks @HawkinsBA for the PR and @simonjayhawkins for review
